### ISSUE TITLE
targets immediately outdated after creation when using `dplyr::select` in command

### ISSUE THREAD

**Issue by mir-cat** [ropensci/drake#503]
I've been experimenting with `drake` for a possible workflow. I have found a confusing behavior though, where it seems some of my targets are outdated immediately after being made. Running `drake::make()` a second time seems to fix it, but I just was wondering what might be causing this behavior. A reprex is below.

``` r
set.seed(2018 - 08 - 15)
library(purrr)
library(dplyr)
#> 
#> Attaching package: 'dplyr'
#> The following objects are masked from 'package:stats':
#> 
#>     filter, lag
#> The following objects are masked from 'package:base':
#> 
#>     intersect, setdiff, setequal, union
library(drake)

# Data --------------------------------------------------------------------
set.seed(2018 - 08 - 15)


fct_var <- function() {
  as.factor(rbinom(100, 1, 0.5))
}

cnt_var <- function() {
  rnorm(100)
}

gen_data <- function() {
  if (rbinom(1, 1, 0.5)) {
    return(fct_var() )
  }

  return(cnt_var())
}


# Setup -------------------------------------------------------------------

plan <-
  drake_plan(
    data = rerun(10, gen_data()) %>%
      bind_cols() %>%
      mutate(y_c = cnt_var(), y_f = fct_var())
  )

make(plan)
#> target data

outdated()
#> character(0)
# Individual Setup --------------------------------------------------------


get_xs <- function(x) {
  grep('^V', names(x), value = T)
}

get_ys <- function(x) {
  grep('^y', names(x), value = T)
}


plan_x <-
  data_frame(target = get_xs(readd('data')),
             command = glue::glue('select(data, {target})')
             )

plan_y <-
  data_frame(target = get_ys(readd('data')),
             command = glue::glue('select(data, {target})')
  )

plan <- bind_plans(plan, plan_x, plan_y)
#> Warning in bind_rows_(x, .id): Vectorizing 'glue' elements may not preserve
#> their attributes
#> Warning in bind_rows_(x, .id): Vectorizing 'glue' elements may not preserve
#> their attributes


outdated()
#> character(0)
make(plan)
#> target V1
#> target V2
#> target V3
#> target V4
#> target V5
#> target V6
#> target V7
#> target V8
#> target V9
#> target V10
#> target y_c
#> target y_f
outdated()
#>  [1] "V1"  "V10" "V2"  "V3"  "V4"  "V5"  "V6"  "V7"  "V8"  "V9"  "y_c"
#> [12] "y_f"
make(plan)
#> target V1
#> target V2
#> target V3
#> target V4
#> target V5
#> target V6
#> target V7
#> target V8
#> target V9
#> target V10
#> target y_c
#> target y_f
outdated()
#> character(0)
```

Created on 2018-08-15 by the [reprex package](http://reprex.tidyverse.org) (v0.2.0).

Additionally, I keep seeing a lot of the error

 #> Warning in is.na(x): is.na() applied to non-(list or vector) of type 'NULL'

though so far there doesn't seem to be any harm that is coming from it.

**Comment by wlandau** [comment]
I may have forgotten to document this, but the loadd() and readd() functions are not meant to be part of commands in your drake_plan(). They are for (1) inspecting cached targets in an interactive session and (2) marking dependencies in knitr reports. I think this could explain the behavior you see.

Let's figure out what your plan should be. What is the goal of this simple example?What would your code look like without drake?

I am at a conference right now, and I will have more time to look at this properly on Friday afternoon.

**Comment by mir-cat** [comment]
The workflow that I am imagining is this: I create a script that runs several hundred models. I was using this using some sort of `mclapply`-like function. Later, I come back to the script, and add a new variable for which I would like to run models. I want to do this without having to rerun all the models that are unaffected by the addition of this new variable. From what I have found, `drake` does this perfectly. Over the course of testing `drake` for this workflow, I happened upon this bug. When I create a function that performs the same functionality as `dplyr::select`, I do not have to build twice for the targets to become current.

Interestingly, it was actually this issue that gave me confidence that `drake` would work for my needs, as it showed me that if a data target was rebuilt and found to be identical, the downstream models that depend on that bit of data wouldn't be rerun.

``` r
library(purrr)
library(dplyr)
#> 
#> Attaching package: 'dplyr'
#> The following objects are masked from 'package:stats':
#> 
#>     filter, lag
#> The following objects are masked from 'package:base':
#> 
#>     intersect, setdiff, setequal, union
library(drake)

# Data --------------------------------------------------------------------
set.seed(2018 - 08 - 16)


fct_var <- function() {
  as.factor(rbinom(100, 1, 0.5))
}

cnt_var <- function() {
  rnorm(100)
}

var <- function() {
  if (rbinom(1, 1, 0.5)) {
    return(fct_var() )
  }

  return(cnt_var())
}

gen_data <- function() {
  rerun(6, var()) %>%
    bind_cols() %>%
    mutate(y_1 = cnt_var(),
           y_2 = cnt_var())
}


# Setup -------------------------------------------------------------------

data <-  gen_data()

get_xs <- function(x) {
  grep('^V', names(x), value = T)
}

get_ys <- function(x) {
  grep('^y', names(x), value = T)
}

lm2 <- function(y, x) {

  dat <- bind_cols(y, x)
  y <- names(dat)[1]

  fo <- glue::glue('{y} ~ .') %>%
    as.formula()

  lm(fo, data = dat)
}


getter <- function(data, x) {
  data[, x]
}

# Drake Plan -------------------------------------------------------------

create_plan <- function(dplyr = TRUE) {
  #browser()
  xs <- get_xs(data)
  ys <- get_ys(data)

  pairs <-
    cross(list(y = ys, x = xs)) %>%
    bind_rows()

  cmnd <- ifelse(dplyr, 'select(data, {target})', 'getter(data, "{target}")')

  setup_plan <-
    data_frame(target = c(xs, ys),
               command = glue::glue(cmnd)
               )

  model_plan <-
    data_frame(
      target = apply(pairs, 1, paste, collapse = '__'),
      command = pmap_chr(pairs, function(y, x) glue::glue('lm2({y}, {x})'))
    )


  bind_plans(setup_plan, model_plan)

}

clean(destroy = TRUE)
plan <- create_plan(dplyr = TRUE)
#> Warning in bind_rows_(x, .id): Vectorizing 'glue' elements may not preserve
#> their attributes

# Need to run make twice for the data targets to be current
make(plan)
#> target V1
#> target V2
#> target V3
#> target V4
#> target V5
#> target V6
#> target y_1
#> target y_2
#> target y_1__V1
#> target y_1__V2
#> target y_1__V3
#> target y_1__V4
#> target y_1__V5
#> target y_1__V6
#> target y_2__V1
#> target y_2__V2
#> target y_2__V3
#> target y_2__V4
#> target y_2__V5
#> target y_2__V6
make(plan)
#> target V1
#> target V2
#> target V3
#> target V4
#> target V5
#> target V6
#> target y_1
#> target y_2
make(plan)
#> All targets are already up to date.

# Don't use dplyr ------------------------------------------------------------

clean(destroy = TRUE)
plan <- create_plan(dplyr = FALSE)
#> Warning in bind_rows_(x, .id): Vectorizing 'glue' elements may not preserve
#> their attributes

# DO NOT Need to run make twice for the data targets to be current
make(plan)
#> target V1
#> target y_1
#> target y_2
#> target V2
#> target V3
#> target V4
#> target V5
#> target V6
#> target y_1__V1
#> target y_2__V1
#> target y_1__V2
#> target y_2__V2
#> target y_1__V3
#> target y_2__V3
#> target y_1__V4
#> target y_2__V4
#> target y_1__V5
#> target y_2__V5
#> target y_1__V6
#> target y_2__V6
make(plan)
#> All targets are already up to date.
```

Created on 2018-08-16 by the [reprex
package](http://reprex.tidyverse.org) (v0.2.0).

**Comment by wlandau** [comment]
## Explanation

In your `dplyr` plan, your target names appear as symbols in the commands themselves.

```r
> plan
# A tibble: 11 x 2
   target command             
   <chr>  <S3: glue>          
 1 mpg    select(mtcars, mpg) 
 2 cyl    select(mtcars, cyl) 
 3 disp   select(mtcars, disp)
...
```

This is a little problematic because `drake::make()` looks for these symbols and treats them as dependencies. For example, in `select(mtcars, mpg)`, `drake` detects symbols `select`, `mtcars`, and `mpg` as the dependencies of `mpg`. The `deps_code()` function shows you exactly which dependencies `drake` is detecting.

```r
> deps_code(quote(select(mtcars, mpg)))
$`globals`
[1] "select" "mtcars" "mpg" 
```

## The bug

Several months ago, in order to solve #216, I forced `drake` to exclude the target name as a dependency. But because of some intense internal refactoring I did recently, it appears the issue has cropped up again. `deps_code()` has no way of knowing about the target name, which is fine, but the new internals need to catch up. I hope to fix this within the next few days. It should not take too long.

## Recommendation

For now, please do not use the target name as a symbol in your commands.

## Short-term development (sufficient to close this issue)

- [ ] Exclude the target name as a dependency in the new internals in `build_drake_graph()`.
- [ ] Document the `loadd()` and `readd()` functions more carefully to explain their proper usage (should not be in commands, should be in interactive sessions and `knitr` reports). 

## Final note

Symbols such as `mpg` in `select(mtcars, mpg)` are tricky. These are called non-standard evaluation (NSE) variables. Some of the time, we can safely ignore these variables, and in your case, we definitely should. However, in cases like #421, we actually need them. Whether to ignore or keep NSEs is a tough decision, and I do not know the right thing to do. It will take careful thought. In the short term, I will keep going with the easiest, simplest, most internally consistent solution: do not treat NSE variables differently in `drake`. But really, the correct solution is to do the thing that is least surprising overall. @gmbecker and I discussed this just yesterday. [`CodeDepends`](https://github.com/duncantl/CodeDepends) may possibly help if we can avoid #421, do code analysis fast, #268, and hopefully avoid the slight friction caused by https://github.com/duncantl/CodeDepends/issues/34).

**Comment by wlandau** [comment]
Fixed in 1cbe3530ce6c7ab6b830706f2209232f3e4ffa70, ea73abeb5eff780f77d3bd643c00057bebb6ebb2, and especially 0dc10a208f396152c4434f2beb3af3fb0b15a0fe.
